### ISSUE TITLE
add a kaldi rule (at least for sleep/wake)

### ISSUE THREAD

**Issue by kendonB** [dictation-toolbox/Caster#797]
**Is your feature request related to a problem? Please describe.**
Dragon has a bunch of built in commands that help with use. It would be nice for switchers to have a Kaldi rule loosely based on the functionality available in base Dragon.

**Describe the solution you'd like**
A Kaldi grammar with the following features:
 - [ ] Sleep/wake

Just the one for now but others may be appropriate especially as KAG adds functionality.

To add sleep/wake we need to translate the grammar in [this file](https://github.com/dictation-toolbox/dragonfly/blob/master/dragonfly/examples/kaldi_module_loader_plus.py) into caster.

@daanzu if you have any pointers or know of someone who has done this please let us know.

**Comment by LexiconCode** [comment]
This potential isn't too hard to implement. This would be relevant to all engines not just Kaldi. How Kaldi  manages to make this work `set_exclusiveness()`. An exclusive grammar takes precedent over all other active grammars. A Rule/grammars that are exclusive are only available for recognition.

Note setting a grammar to exclusive overrides DNS's built-in sleep-wake function. When using DNS "natlink.setMicState("sleeping")" setting to match the state of the grammar.

The FunkContext with a mapping rule could be used when sleep state is enabled. The enabling command could be housed within `utility_rules\caster_rule.py`

**Comment by lexxish** [comment]
> This potential isn't too hard to implement. This would be relevant to all engines not just Kaldi. How Kaldi manages to make this work `set_exclusiveness()`. An exclusive grammar takes precedent over all other active grammars. A Rule/grammars that are exclusive are only available for recognition.
> 
> Note setting a grammar to exclusive overrides DNS's built-in sleep-wake function. When using DNS "natlink.setMicState("sleeping")" setting to match the state of the grammar.
> 
> The FunkContext with a mapping rule.

Overriding the DNS default commands for sleep/wake would be nice because you can use shorter commands such as "snore" to put the mic to sleep.

Kaldi implementation is more important in my opinion but, to use this idea for Natlink it can be done with something like:
```
def snore():
    import natlink
    engine = get_engine()
    engine.speak("Sleeping")
    natlink.setMicState("sleeping")

def wake():
    import natlink
    natlink.setMicState("on")
    engine = get_engine()
    engine.speak("Awake")
```

**Comment by LexiconCode** [comment]
> Only downside is I'm not sure if we could get the taskbar icon toggling from green to blue. 

Fortunately I believe this can be handled by `natlink.setMicState(state)` reading the documentation it controls the mic, where state is 'on', 'off' or 'sleeping'
natlink.getMicState() returns current state. Therefore the DNS icon could be in sync with the exclusive grammar state.

**Comment by lexxish** [comment]
> > Only downside is I'm not sure if we could get the taskbar icon toggling from green to blue.
> 
> Fortunately I believe this can be handled by `natlink.setMicState(state)` reading the documentation it controls the mic, where state is 'on', 'off' or 'sleeping' natlink.getMicState() returns current state. Therefore the DNS icon could be in sync with the exclusive grammar state.
> 
> I agree the though the implementation is more important for WSR/Kaldi.

Regarding Kaldi, would implementation involve changing https://github.com/dictation-toolbox/Caster/blob/master/castervoice/lib/ctrl/mgr/loading/load/content_loader.py or does this rule operate independent of how you load the other rules?

**Comment by LexiconCode** [comment]
fortunately we don't have to change anything in Caster to make grammars exclusive. It's a simple bool. It works on any other rule. The rule must be already loaded into the engine before it's set to be exclusive. One set no other commands except for those that are exclusive through one or multiple rules will be recognized.

**Comment by kendonB** [comment]
@lexxish did you ever figure out getting sleep to work?

**Comment by LexiconCode** [comment]
With straight dragonfly this would be pretty easy with Caster it's a bit different because we don't know the grammar name being used as it's different every boot. I've been working on programmatically switching DNS Modes in preparation for creating a mode unified mode manager for all engines. The following could be used in the sleep grammar.

```
def find_grammar_name():
    grammar_cach = None
    if grammar_cach is None:
        for grammar in get_engine().grammars:
            for rule in grammar.rules:
                if rule.exported:
                    if rule.name == "Mode Rules": # Edit for the rule name which contains grammar (non-CCR mappingrule only)
                        print(rule.name)
                        grammar_cach = grammar
                        return grammar_cach
    else:               
        return grammar_cach
```
in another function then you can use `grammar_cach.set_exclusiveness(0)` or `grammar_cach.set_exclusiveness(1)` to toggle exclusiveness

**Comment by LexiconCode** [comment]
>     * A "maximize window" command that works correctly.

What's wrong with the current behavior @kendonB?

> An automatic program switcher. In Dragon. you say "switch to ".
Besides creating a GUI the backend information could be obtained from a tweaked function to use `get_all_windows():` returning all pids list instead of `Window.get_foreground()`

```
def get_active_window_info():
    '''Returns foreground window executable_file, executable_path, title, handle, classname'''
    FILENAME_PATTERN = re.compile(r"[/\\]([\w_ ]+\.[\w]+)")
    window = Window.get_foreground()
    executable_path = str(Path(get_active_window_path()))
    match_object = FILENAME_PATTERN.findall(window.executable)
    executable_file = None
    if len(match_object) > 0:
        executable_file = match_object[0]
    return [executable_file, executable_path, window.title, window.handle, window.classname]
```

**Comment by lexxish** [comment]
> @lexxish did you ever figure out getting sleep to work?

I have not tried yet. Will update you all if I do. 

I do have some "switch to" like code I can post if anyone wants it. I use a phonetic distance library to choose the best match based on what is currently running. Also have "open" like command that searches a couple directories (e.g. desktop)...it's not perfect and I think the way "bring" allows you to specify programs is also nice for things you use a lot.

Another item that would be nice would be ability to use Kaldi for commands, but DNS for dictation - similar to how I believe Kaldi can be used with Google Speech Recognition. 

Last item that would be nice to have (but deserves it's own issue number) is integration with accessibility APIs like DNS has. So you can say things like "Click X" when X is a button in a browser.

**Comment by lexxish** [comment]
> > ```
> > * A "maximize window" command that works correctly.
> > ```
> 
> What's wrong with the current behavior @kendonB?
> 
> > An automatic program switcher. In Dragon. you say "switch to ".
> 
> Besides creating a GUI the backend information could be obtained from a tweaked function to use `get_all_windows():` returning all pids list instead of `Window.get_foreground()`
> 
> ```
> def get_active_window_info():
>     '''Returns foreground window executable_file, executable_path, title, handle, classname'''
>     FILENAME_PATTERN = re.compile(r"[/\\]([\w_ ]+\.[\w]+)")
>     window = Window.get_foreground()
>     executable_path = str(Path(get_active_window_path()))
>     match_object = FILENAME_PATTERN.findall(window.executable)
>     executable_file = None
>     if len(match_object) > 0:
>         executable_file = match_object[0]
>     return [executable_file, executable_path, window.title, window.handle, window.classname]
> ```

I could be wrong, but I think Caster's default maximize uses "alt+SPACE, x" to maximize rather then sending the foreground window a maximize message (https://docs.microsoft.com/en-us/windows/win32/learnwin32/window-messages). I don't think "alt+SPACE, x" works for every application, but can't think of a specific one right now. The same type of scenario for exists for closing windows in Caster too I believe, where we could send SIGTERM and/or SIGKILL message equivalents (probably two different voice commands) instead of using keyboard shortcuts and it would (hopefully) work more consistently.

**Comment by LexiconCode** [comment]
> I could be wrong, but I think Caster's default maximize uses "alt+SPACE, x" to maximize rather then sending the foreground window a maximize message

Back when implementing kaldi support I switched it from `"alt+SPACE, x" ` to dragonfly Win32 events. If something's not behaving correctly with those minimize/maximize commands let me know.

https://github.com/dictation-toolbox/Caster/blob/7d3834eed076d39db1f163582d4e457ab71ee5f4/castervoice/rules/core/navigation_rules/window_mgmt_rule.py#L13

**Comment by LexiconCode** [comment]
> Last item that would be nice to have (but deserves it's own issue number) is integration with accessibility APIs like DNS has. So you can say things like "Click X" when X is a button in a browser.

I agree, I've been looking into the following as many of the UI automation platforms are too slow for our purposes.

https://github.com/yinkaisheng/Python-UIAutomation-for-Windows/blob/master/uiautomation/uiautomation.py

The [UIAutomationClient](https://github.com/yinkaisheng/UIAutomationClient) does not include Cache and Events yet which is unfortunate.

The best tool to date to explore applications to see what's accessible for the Microsoft's UIautomation is [Accessibility Insights for Windows](https://accessibilityinsights.io/docs/en/windows/overview)

**Comment by daanzu** [comment]
> Another item that would be nice would be ability to use Kaldi for commands, but DNS for dictation - similar to how I believe Kaldi can be used with Google Speech Recognition.

I don't have experience with Natlink, and don't currently have Dragon installed, but I'd be happy to help implementing this. Is there a way with Natlink to just get straight dictation recognition text from audio data passed to it? Perhaps there should be an issue in KaldiAG for working on this?

**Comment by LexiconCode** [comment]
> Perhaps there should be an issue in KaldiAG for working on this?
Agreed

**Comment by LexiconCode** [comment]
@lexxish and @kendonB I will attempt to implement the sleeping grammar and modes for all engines. These modes will override DNS's built-in modes but will be kept in sync with the DNS GUI.

**Comment by LexiconCode** [comment]
The https://github.com/dictation-toolbox/Caster/pull/881 addresses the following request.

> An automatic program switcher. In Dragon. you say "switch to ".
